### PR TITLE
Improvement: Revamp TOC with dynamic logic.

### PR DESCRIPTION
**Pull request by AbhijitNage123**
### Description

- https://trello.com/c/604tdcVT/172-revamp-toc-with-dynamic-logic

### Types of changes

- New feature (non-breaking change which adds functionality)

### How has this been tested?

- Drag and drop TOC and check every possible case with heading

### Checklist:
- [x] My code is tested
- [ ] My code follows accessibility standards 
- [ ] My code has proper inline documentation 
- [x] I've included any necessary tests 
- [ ] I've included developer documentation 
- [x] I've added proper labels to this pull request

### CODE REVIEW

File: dist/blocks/table-of-content/class-uagb-table-of-content.php [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+								'type' => "string",
+							),
+							'align' => array(
+								'type'    => "string",
+								'default' => "left",
+							),
+							'heading'=> array(
+								'source'=> "html",
+								'selector'=> ".uagb-toc__title",
+								'default'=> "Table Of Contents",
```
Comment (amithire1201) [inline dist/blocks/table-of-content/class-uagb-table-of-content.php:105]: Can we make this string translatable?

Comment (AbhijitNage123) [reply dist/blocks/table-of-content/class-uagb-table-of-content.php]: Done

**Review by amithire1201 — APPROVED** [review]

File: dist/blocks/table-of-content/class-uagb-table-of-content.php [hunk 2]
```diff
@@ -0,0 +1,620 @@
...
+
+			$blocks         = parse_blocks( $post->post_content );
+			$level          = '';
+			$tag            = '';
+			$excludeheading = '';
+
+			foreach ( $blocks as $block ) {
+				$innercount = count( $block['innerBlocks'] );
+				if ( 'core/heading' === $block['blockName'] ) {
+
```
Comment (RajkiranBagal) [inline dist/blocks/table-of-content/class-uagb-table-of-content.php:394]: Remove the repetitive code. Make it like this.
```Remove the repetitive code.
if ( 'core/heading' === $block['blockName'] || 'uagb/advanced-heading' === $block['blockName'] ) {

					$tag            = ( isset( $block['attrs']['level'] ) ) ? $block['attrs']['level'] : 2;
					$level          = ( isset( $block['attrs']['level'] ) ) ? $block['attrs']['level'] : 2;  // h2 as default.
					$excludeheading = ( isset( $block['attrs']['className'] ) ) ? $block['attrs']['className'] : '';

					$headings[] = array(
						'tag'            => $tag,
						'title'          => wp_strip_all_tags( $block['innerHTML'] ),
						'link'           => strtolower( $this->remove_special_char( wp_strip_all_tags( $block['innerHTML'] ) ) ),
						'level'          => ( $level ),
						'excludeheading' => $excludeheading,
					);
				}
```

File: dist/blocks/table-of-content/class-uagb-table-of-content.php [hunk 3]
```diff
@@ -0,0 +1,620 @@
...
+					$tag            = ( isset( $block['attrs']['level'] ) ) ? $block['attrs']['level'] : 2;
+					$level          = ( isset( $block['attrs']['level'] ) ) ? $block['attrs']['level'] : 2;  // h2 as default.
+					$excludeheading = ( isset( $block['attrs']['className'] ) ) ? $block['attrs']['className'] : '';
+
+					$headings[] = array(
+						'tag'            => $tag,
+						'title'          => wp_strip_all_tags( $block['innerHTML'] ),
```
Comment (RajkiranBagal) [inline dist/blocks/table-of-content/class-uagb-table-of-content.php:401]: Do the same thing as suggested in the above comment in the below conditions.

**Review by RajkiranBagal — CHANGES_REQUESTED** [review]

File: dist/blocks/table-of-content/class-uagb-table-of-content.php [hunk 4]
```diff
@@ -0,0 +1,582 @@
...
+
+		/**
+		 * Filter Heading array
+		 *
+		 * @param  array $filteredHeaders variable.
+		 *
+		 * @since x.x.x
+		 */
+		public function filter_array( $filteredHeaders ) {
+			$sortedHeaders = array();
```
Comment (RajkiranBagal) [inline dist/blocks/table-of-content/class-uagb-table-of-content.php:490]: Why are we using camelCase variable names in PHP? Let's keep them consistent, throughout the Plugin.

ex: `$sortedHeaders` => `$sorted_headers`

Do this change for all such variables in this file.

**Review by RajkiranBagal — CHANGES_REQUESTED** [review]

File: dist/blocks/table-of-content/class-uagb-table-of-content.php (hunk 2 above)
Comment (AbhijitNage123) [reply dist/blocks/table-of-content/class-uagb-table-of-content.php:394]: Resolved.

File: dist/blocks/table-of-content/class-uagb-table-of-content.php (hunk 3 above)
Comment (AbhijitNage123) [reply dist/blocks/table-of-content/class-uagb-table-of-content.php]: Resolved.

File: dist/blocks/table-of-content/class-uagb-table-of-content.php (hunk 4 above)
Comment (AbhijitNage123) [reply dist/blocks/table-of-content/class-uagb-table-of-content.php]: Resolved for all

**Review by RajkiranBagal — APPROVED** [review]

File: assets/js/table-of-contents.js [hunk 5]
```diff
@@ -129,22 +126,75 @@
 			}
 
 			var all_header = ( undefined !== allowed_h_tags_str && '' !== allowed_h_tags_str ) ? $( 'body' ).find( allowed_h_tags_str ) : $( 'body' ).find('h1, h2, h3, h4, h5, h6' );
+			var TOC = '';
```
Comment (RajkiranBagal) [inline assets/js/table-of-contents.js:129]: Add proper variable names in camelCase

File: assets/js/table-of-contents.js [hunk 6]
```diff
@@ -129,22 +126,75 @@
...
+				all_header.each( function (){
+					let header = $( this );
+					let header_text = parseTocSlug(header.text());
+					$( this ).before('<span id="'+ header_text +'" class="uag-toc__heading-anchor"></span>');					
+				});				
+			}
 
-			if ( undefined !== $headers && 0 !== all_header.length ) {
+			var header_array = $( 'div.entry-content' ).find( all_header )
+			console.log(header_array)
```
Comment (RajkiranBagal) [inline assets/js/table-of-contents.js:142]: Remove console.logs

File: assets/js/table-of-contents.js [hunk 7]
```diff
@@ -129,22 +126,75 @@
...
-						let header = $( this );
-						let header_text = parseTocSlug(header.text());
+				if ( !exclude_heading ) {
+					
+					let openLevel = header[0].localName.replace(/^h+/, '');
+					let titleText = header.text();
+					
+					if (openLevel > level) {
+						
+						if( 2 == (new Array(openLevel - level + 1)).length ){
```
Comment (RajkiranBagal) [inline assets/js/table-of-contents.js:165]: Why are we declaring new Array in conditions & making calculations?

File: assets/js/table-of-contents.js [hunk 8]
```diff
@@ -129,22 +126,75 @@
...
+				}					
+			});
+
+			$(".uagb_table-of-contents-placeholder").remove();
+
+			$(".uagb-toc__list-wrap").prepend(TOC);
+
+			} else{
+
+				TOC +=  'Add a header to begin generating the table of contents';
```
Comment (RajkiranBagal) [inline assets/js/table-of-contents.js:192]: This should be translatable.

File: classes/class-uagb-config.php [hunk 9]
```diff
@@ -48,6 +48,7 @@ public static function get_block_attributes() {
 						'default'     => true,
 						'attributes'  => array(
 							'classMigrate'            => false,
+							'headingTitleString'      => '',
```
Comment (RajkiranBagal) [inline classes/class-uagb-config.php:51]: Remove this if we are not using

**Review by RajkiranBagal — CHANGES_REQUESTED** [review]

**Comment by RajkiranBagal** [comment]
@AbhijitNage123 Pull latest next-release.

File: assets/js/table-of-contents.js (hunk 5 above)
Comment (AbhijitNage123) [reply assets/js/table-of-contents.js]: Updated

File: assets/js/table-of-contents.js (hunk 6 above)
Comment (AbhijitNage123) [reply assets/js/table-of-contents.js]: I need it for debugging. I'll remove it later

File: assets/js/table-of-contents.js (hunk 7 above)
Comment (AbhijitNage123) [reply assets/js/table-of-contents.js]: Resolved

File: assets/js/table-of-contents.js (hunk 8 above)
Comment (AbhijitNage123) [reply assets/js/table-of-contents.js]: Resolved

File: classes/class-uagb-config.php (hunk 9 above)
Comment (AbhijitNage123) [reply classes/class-uagb-config.php]: Resolved

File: classes/class-uagb-config.php [hunk 10]
```diff
@@ -38,8 +38,7 @@ class UAGB_Config {
 		 * @return array The Widget List.
 		 */
 		public static function get_block_attributes() {
-			$keys = array(0,1,2,3,4,5);
-			$array = array_fill_keys($keys, true);
+			$array = array_fill_keys(array(0,1,2,3,4,5), true);
```
Comment (RajkiranBagal) [inline classes/class-uagb-config.php:41]: Please add a meaningful name to the array. Also, add proper indentation.

File: assets/js/table-of-contents.js [hunk 11]
```diff
@@ -117,34 +116,87 @@
...
 			if ( undefined !== attr.mappingHeaders ) {
 
 				attr.mappingHeaders.forEach(function(h_tag, index) { (h_tag === true ? allowed_h_tags.push('h' + (index+1)) : null);});
 				var allowed_h_tags_str = ( null !== allowed_h_tags ) ? allowed_h_tags.join( ',' ) : '';
+
+				headerMappingHeaders = attr.mappingHeaders.filter(header => header).length;
 			}
 
 			var all_header = ( undefined !== allowed_h_tags_str && '' !== allowed_h_tags_str ) ? $( 'body' ).find( allowed_h_tags_str ) : $( 'body' ).find('h1, h2, h3, h4, h5, h6' );
+			var UAGBToc = '';
```
Comment (RajkiranBagal) [inline assets/js/table-of-contents.js:131]: This is not camelCase @AbhijitNage123 .

File: dist/blocks/table-of-content/class-uagb-table-of-content.php [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+		 *
+		 * @since x.x.x
+		 */
+		public function register_blocks() {
+			// Check if the register function exists.
+			if ( ! function_exists( 'register_block_type' ) ) {
+				return;
+			}
+
+			$array = array_fill_keys(array(0,1,2,3,4,5), true);
```
Comment (RajkiranBagal) [inline dist/blocks/table-of-content/class-uagb-table-of-content.php:60]: Add meaningful array name.

File: src/blocks/table-of-contents/attributes.js [hunk 13]
```diff
@@ -69,7 +69,7 @@ const attributes = {
 		default : "left",
 	},
 	heading: {
-		source: "html",
+		type: "string",
 		selector: ".uagb-toc__title",
 		default: "Table Of Contents",
```
Comment (RajkiranBagal) [inline src/blocks/table-of-contents/attributes.js:74]: Add Translation.

File: src/blocks/table-of-contents/edit.js [hunk 14]
```diff
@@ -1060,42 +1019,43 @@ export default compose(
...
+					exclude_heading = true;
+				} else if ( 0 < header.parents('.uagb-toc-hide-heading').length ) {
+					exclude_heading = true;
+				} else {
+					exclude_heading = false;
 				}
-
-					if ( !headingContentEmpty && !exclude_heading ) {
+				
+				let header_text = parseTocSlug(header.text());
```
Comment (RajkiranBagal) [inline src/blocks/table-of-contents/edit.js:1040]: Please keep consistent variables, Some are camelCase some are another format. Update them accordingly.

**Review by RajkiranBagal — CHANGES_REQUESTED** [review]

File: classes/class-uagb-config.php (hunk 10 above)
Comment (AbhijitNage123) [reply classes/class-uagb-config.php]: Updated.

File: assets/js/table-of-contents.js (hunk 11 above)
Comment (AbhijitNage123) [reply assets/js/table-of-contents.js]: Updated.

File: dist/blocks/table-of-content/class-uagb-table-of-content.php (hunk 12 above)
Comment (AbhijitNage123) [reply dist/blocks/table-of-content/class-uagb-table-of-content.php]: Added

File: src/blocks/table-of-contents/attributes.js (hunk 13 above)
Comment (AbhijitNage123) [reply src/blocks/table-of-contents/attributes.js]: Done

File: src/blocks/table-of-contents/edit.js (hunk 14 above)
Comment (AbhijitNage123) [reply src/blocks/table-of-contents/edit.js]: Done

File: assets/js/table-of-contents.js [hunk 15]
```diff
@@ -117,34 +116,87 @@
...
+	
+				all_header.each( function (){
+					let header = $( this );
+					let header_text = parseTocSlug(header.text());
+					$( this ).before('<span id="'+ header_text +'" class="uag-toc__heading-anchor"></span>');					
+				});				
+			}
+			let blockId = attr.block_id;
+			var headerArray = $( 'div.entry-content' ).find( all_header )
+			console.log(headerArray) // For testing will remove it later
```
Comment (vrundakansara) [inline assets/js/table-of-contents.js:144]: Remove debugging logs.

File: classes/class-uagb-config.php [hunk 16]
```diff
@@ -38,7 +38,7 @@ class UAGB_Config {
 		 * @return array The Widget List.
 		 */
 		public static function get_block_attributes() {
-
+			$mappingHeadersArray = array_fill_keys(array(0,1,2,3,4,5), true);
```
Comment (vrundakansara) [inline classes/class-uagb-config.php:41]: Why do we need this? Can we declare this statically where it is needed?

File: classes/class-uagb-config.php [hunk 17]
```diff
@@ -2261,6 +2261,7 @@ public static function get_block_attributes() {
 							'classMigrate'              => false,
 							'block_id'                  => '',
 							'heading'                   => 'Table Of Contents',
+							'emptyHeadingTeaxt'         => 'Add a header to begin generating the table of contents',
```
Comment (vrundakansara) [inline classes/class-uagb-config.php:2264]: Make this string translatable.

**Review by vrundakansara — CHANGES_REQUESTED** [review]

**Review by vrundakansara — CHANGES_REQUESTED** [review]

File: dist/blocks/table-of-content/class-uagb-table-of-content.php [hunk 18]
```diff
@@ -0,0 +1,435 @@
...
+		}
+
+		/**
+		 * Renders the Table of content block on server.
+		 *
+		 * @param array $attributes Array of block attributes.
+		 *
+		 * @since x.x.x
+		 */
+		public function uagb_render_table_of_contents_block( $attributes ) {
```
[the remark this answers is not in the document]
Comment (AbhijitNage123) [reply dist/blocks/table-of-content/class-uagb-table-of-content.php]: Done

File: dist/blocks/table-of-content/class-uagb-table-of-content.php [hunk 19]
```diff
@@ -0,0 +1,435 @@
...
+
+			$wrap = array(
+				'wp-block-uagb-table-of-contents ',
```
[the remark this answers is not in the document]
Comment (AbhijitNage123) [reply dist/blocks/table-of-content/class-uagb-table-of-content.php:393]: Not working. Added support of Additional CSS class(es) in PHP file

**Review by vrundakansara — APPROVED** [review]

**Pull request merged**